File: src/main/java/dev/rvbsm/fsit/api/event/PassedUseBlockCallback.java
package dev.rvbsm.fsit.api.event;

import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.ActionResult;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.world.World;
import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;

import org.jetbrains.annotations.NotNull;

@FunctionalInterface
public interface PassedUseBlockCallback {

    Event<PassedUseBlockCallback> EVENT = EventFactory.createArrayBacked(
        PassedUseBlockCallback.class, (listeners) -> (player, world, hitResult) -> {
            for (PassedUseBlockCallback listener : listeners) {
                final ActionResult result = listener.interact(player, world, hitResult);

                if (result != ActionResult.PASS) {
                    return result;
                }
            }

            return ActionResult.PASS;
        });

    ActionResult interact(@NotNull ServerPlayerEntity player, @NotNull World world, @NotNull BlockHitResult hitResult);
}

File: src/main/kotlin/dev/rvbsm/fsit/event/PassedUseBlockCallback.kt
package dev.rvbsm.fsit.event

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.ActionResult
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.world.World

fun interface PassedUseBlockCallback {
    fun interactBlock(player: ServerPlayerEntity, world: World, hitResult: BlockHitResult): ActionResult

    companion object {
        @JvmField
        val EVENT: Event<PassedUseBlockCallback> =
            EventFactory.createArrayBacked(PassedUseBlockCallback::class.java) { listeners ->
                PassedUseBlockCallback { player, world, hitResult ->
                    for (listener in listeners) {
                        val result = listener.interactBlock(player, world, hitResult)

                        if (result != ActionResult.PASS) {
                            return@PassedUseBlockCallback result
                        }
                    }

                    return@PassedUseBlockCallback ActionResult.PASS
                }
            }
    }
}
